docs/src/manual/compiling_lux_models.md

```diff
@@ -0,0 +1,171 @@
+> functions will assume the same control flow pattern as was original taken by objects used
+> at compile time, and control flow (e.g. if, for) as well as any type instabilities will be
+> removed. The benefits of this approach is immediately making all such code available for
+> advanced optimization with little developer effort.
+
+!!! danger "Experimental"
+
+    Reactant compilation is a very new feature and is currently experimental. Certain models
+    might not be compilable yet, but we are actively working on it. Open an issue if you
+    encounter any problems.
+
+```@example compile_lux_model
+using Lux, Reactant, Enzyme, Random, Zygote
+using Functors, Optimisers, Printf
+```
+
+!!! tip "Using the `TrainState` API"
+
+    If you are using the [`Training.TrainState`](@ref) API, skip to the
+    [bottom of this page](@ref compile_lux_model_trainstate) to see how to train the model
+    without any of this boilerplate.
+
+We start by defining a simple MLP model:
+
+```@example compile_lux_model
+model = Chain(
+    Dense(2 => 32, gelu),
+    Dense(32 => 32, gelu),
+    Dense(32 => 2)
+)
+ps, st = Lux.setup(Random.default_rng(), model)
+```
+
+We then create a random input and output data:
+
+```@example compile_lux_model
+x = randn(Float32, 2, 32)
+y = x .^ 2
+```
+
+We will use [`xla_device`](@ref) similar to [`gpu_device`](@ref) to move the arrays to
+`Reactant`.
+
+```@example compile_lux_model
+const xdev = xla_device()
+
+x_ra = x |> xdev
+y_ra = y |> xdev
+ps_ra = ps |> xdev
+st_ra = st |> xdev
+nothing
+```
+
+First let's run the model as we would normally:
+
+```@example compile_lux_model
+pred_lux, _ = model(x, ps, Lux.testmode(st))
+```
+
+To run it using `XLA` we need to compile the model. We can do this using the
+`Reactant.@compile` macro. Note that the inputs need to be moved to the device using
+[`xla_device`](@ref) first.
+
+```@example compile_lux_model
+model_compiled = @compile model(x_ra, ps_ra, Lux.testmode(st_ra))
+```
+
+Now we can test the difference between the results:
+
+```@example compile_lux_model
+pred_compiled, _ = model_compiled(x_ra, ps_ra, Lux.testmode(st_ra))
+
+pred_lux .- Array(pred_compiled)
+```
+
+The difference is very small as we would expect. Now, let's try to differentiate the
+output of the model. We need to use `Enzyme.jl` to do this.
+
+```@example compile_lux_model
+function loss_function(model, ps, st, x, y)
+    pred, _ = model(x, ps, st)
+    return MSELoss()(pred, y)
+end
+```
+
+We will use `Zygote.jl` to compute the gradient of the loss function for the vanilla model.
+
+```@example compile_lux_model
+loss_function(model, ps, st, x, y)
+
+∂ps_zyg = only(Zygote.gradient(ps -> loss_function(model, ps, st, x, y), ps))
+```
+
+Now we will compile the gradient function using `Reactant.@compile`.
+
+```@example compile_lux_model
+function enzyme_gradient(model, ps, st, x, y)
+    dps = Enzyme.make_zero(ps)
+    Enzyme.autodiff(Enzyme.Reverse, Const(loss_function), Active, Const(model),
```

Comment:
if dps isn't preallocated there's now some nicer syntax of 

```julia
Enzyme.gradient(Reverse, Const(loss_function), Const(model), ps, Const(st), Const(x), Const(y))
```

Reply:
BTW TPUs also working nicely https://colab.research.google.com/drive/1kVP_wLI4vFhR32ggImdssCj0xI7XR00u?usp=sharing